Source/Scene/DebugInspector.js

```diff
@@ -0,0 +1,155 @@
+import clone from "../Core/clone.js";
+import Color from "../Core/Color.js";
+import DrawCommand from "../Renderer/DrawCommand.js";
+import ShaderSource from "../Renderer/ShaderSource.js";
+import ShaderProgram from "../Renderer/ShaderProgram.js";
+import defined from "../Core/defined.js";
+
+/**
+ * @private
+ */
+function DebugInspector() {
+  this._cachedShowFrustumsShaders = {};
+}
+
+function getAttributeLocations(shaderProgram) {
+  var attributeLocations = {};
+  var attributes = shaderProgram.vertexAttributes;
+  for (var a in attributes) {
+    if (attributes.hasOwnProperty(a)) {
+      attributeLocations[a] = attributes[a].index;
+    }
+  }
+
+  return attributeLocations;
+}
+
+function createDebugShowFrustumsShaderProgram(scene, shaderProgram) {
+  var context = scene.context;
+  var sp = shaderProgram;
+  var fs = sp.fragmentShaderSource.clone();
+
+  var targets = [];
+  fs.sources = fs.sources.map(function (source) {
+    source = ShaderSource.replaceMain(source, "czm_Debug_main");
+    var re = /gl_FragData\[(\d+)\]/g;
+    var match;
+    while ((match = re.exec(source)) !== null) {
+      if (targets.indexOf(match[1]) === -1) {
+        targets.push(match[1]);
+      }
+    }
+    return source;
+  });
+  var length = targets.length;
+
+  var newMain = "";
+  newMain += "uniform vec3 debugShowCommandsColor;\n";
+  newMain += "uniform vec3 debugShowFrustumsColor;\n";
+  newMain += "void main() \n" + "{ \n" + "    czm_Debug_main(); \n";
+
+  // set debugShowCommandsColor to Color(1.0, 1.0, 1.0, 1.0) to stop rendering scene.debugShowCommands
+  // set debugShowFrustumsColor to Color(1.0, 1.0, 1.0, 1.0) to stop rendering scene.debugShowFrustums
+  var i;
+  if (length > 0) {
+    for (i = 0; i < length; ++i) {
+      newMain +=
+        "    gl_FragData[" + targets[i] + "].rgb *= debugShowCommandsColor;\n";
+      newMain +=
+        "    gl_FragData[" + targets[i] + "].rgb *= debugShowFrustumsColor;\n";
+    }
+  } else {
+    newMain += "    gl_FragColor.rgb *= debugShowCommandsColor;\n";
+    newMain += "    gl_FragColor.rgb *= debugShowFrustumsColor;\n";
+  }
+  newMain += "}";
+
+  fs.sources.push(newMain);
+
+  var attributeLocations = getAttributeLocations(sp);
+
+  return ShaderProgram.fromCache({
+    context: context,
+    vertexShaderSource: sp.vertexShaderSource,
+    fragmentShaderSource: fs,
+    attributeLocations: attributeLocations,
+  });
+}
+
+var scratchFrustumColor = new Color();
+function createDebugShowFrustumsUniformMap(scene, command) {
+  // setup uniform for the shader
+  var debugUniformMap;
+  if (!defined(command.uniformMap)) {
+    debugUniformMap = {};
+  } else {
+    debugUniformMap = clone(command.uniformMap);
+    // debugUniformMap = command.uniformMap;
```

Comment:
oh I left a github comment here, but it maybe not shown up to you. My first time using github to contribute to the open source, so I'm still confused about some of its functionalities. Let me paste the comment below and the [Sandcastle example](http://localhost:8080/Apps/Sandcastle/index.html#c=xVXbThsxEP0VK08bKXWgIQqkATUJUFWCgiDQPqyEnHjYWPXaK9sbEqr8e2e92QshtC+95GHjyzlnLp6xF8yQhYAnMOSYKHgiY7Aijem9XwvCxszPx1o5JhSYsNEiP0JFiANjcOXa6IXgYPoFcWaAOfiqjeSTHBI0W6FaNz+EKlQLNGdnoACt5Wapn+Km/6eR1FOgHBI3n4B1wwgFrNsoIcmZtAJzmKbR7Vw/jXUcM8VtBQhVuz3kfOMV+axsAjOnDbqQm4WlA8WDjdf5Yj4psZdiKVSzJnZtRCycWIAtPEjKFco4D7LE1JJY4gOfMkIi0DE4s0ITjilMbb8O/7S1W7AqXpnlkV4WaPpodHwqYlBWaGUrEiELMA6W59rEzJXUazCFgbGW2gyTBJjJpvT+7GZy9u3h/OrmcjhpVTq8VH/h75ihvBVMdYKDvexH91qkU466m1GzVFpXw1hzkJfMGbEsPcunBzROpROJXI1WE/TLSubQdFC5s4H7zUcMzlJg1n3Rxs3vkok+F0vg54bFUOOQ107niYPIANjg3f5eN3P6AL/NitasZWF35D5U/3nvA36DsYktKLUrHHO4NU1dVg21w5tlh1Pmxh/VdoEMC6YPxUN2hpxt0JuzU/ok3HwokzlDt7s7w1xvRuVRsbI8Xpz921VUFeBMagu871vypewabeeN9Z/aqAa/wV5nKpJQ8OodZIrN/iv4i/KpZx1LKauH+lJne+HdUWcb0qN7u+tuDiKaY//u+35q/en2Xv/zShxd3J39pdrbLq/svZkKKaeaGf887Ky3mvaoAKOyxKPObp5mplWpeIpPT6KtyBD9310uvS7tHvV6PbxfsBI6h4edFtmvXRciZhHGGDYobfuxbeeKDxc60g8aT1qyFU1UhO9v+Zg2Wo2BdSsJJ0UGP4o4wWuQpEYGKOUgTvD2RLVpOvsOjs6szagZdNCuUwdcLIjgxzvee2xjZi3uPKZS3opnCBsngzbiX1GlZlyo6Cr3NoPN908u8kVK6aCN091Mp7WcMjRWRlLby9/+S1DpluU66lnreJQ6h2/UL1CWYUJgN64+/gk) for the issue as well:

I try to insert debug uniforms to the current command uniform and stop creating debug uniform closures if they already exist in the command uniform. But there is a problem that when the debug uniform closure captures the command, it may not be the command that is currently being executed. If you un-comment line 87 and have scene.debugShowCommand = true, the cesium billboard's debug color stop changing frequently. I'm not entirely sure, but I think the debug uniform closure only captures one of the billboard's command, which is not correct.

Reply:
Weird, I didn't see your comment. Maybe it was part of an in-progress review that didn't get submitted?

> But there is a problem that when the debug uniform closure captures the command, it may not be the command that is currently being executed

Which suggests the billboard system might be creating a new command every frame but using the same uniform map object. That's interesting...

I prefer the non flickery behavior actually, even though it's different than master. I'd be happy to merge the `debugUniformMap = command.uniformMap` version.